pkg/operator/controller/ingress/router_status.go

```diff
@@ -0,0 +1,198 @@
+//    - The ingress controller knows when a router is deleted because it is the one responsible for deleting it. So it
+//      simply calls clearRouteStatus to clear status of routes that openshift-router has admitted.
+// #2 When the ingress controller sharding configuration (i.e., selectors) is changed.
+//    - When the selectors (routeSelector and namespaceSelector) are updated, the operator simply clears the status of
+//      any route that it is no longer selecting using the updated selectors.
+//    - We determine what routes are admitted by the current state of the selectors (just like the openshift-router).
+
+// syncRouteStatus ensures that all routes status have been synced with the ingress controller's state.
+func (r *reconciler) syncRouteStatus(ic *operatorv1.IngressController) []error {
+	// Clear routes that are not admitted by this ingress controller if route selectors have been updated.
+	if routeSelectorsUpdated(ic) {
+		// Only clear once we are done rolling out routers.
+		// We want to avoid race condition in which we clear status and an old router re-admits it before terminated.
+		if done, err := r.isRouterDeploymentRolloutComplete(ic); err != nil {
+			return []error{err}
+		} else if done {
+			// Clear routes status not admitted by this ingress controller.
+			if errs := r.clearRoutesNotAdmittedByIngress(ic); len(errs) > 0 {
+				return errs
+			}
+
+			// Now sync the selectors from the spec to the status, so we indicate we are done clearing status.
+			if err := r.syncIngressControllerSelectorStatus(ic); err != nil {
+				return []error{err}
+			}
+		}
+	}
+	return nil
+}
+
+// isRouterDeploymentRolloutComplete determines whether the rollout of the ingress router deployment is complete.
+func (r *reconciler) isRouterDeploymentRolloutComplete(ic *operatorv1.IngressController) (bool, error) {
+	deployment := appsv1.Deployment{}
+	deploymentName := operatorcontroller.RouterDeploymentName(ic)
+	if err := r.cache.Get(context.TODO(), deploymentName, &deployment); err != nil {
+		return false, fmt.Errorf("failed to get deployment %s: %w", deploymentName, err)
+	}
+
+	if deployment.Status.Replicas != deployment.Status.UpdatedReplicas {
+		return false, nil
+	}
+	return true, nil
+}
+
+// clearAllRoutesStatusForIngressController clears any route status that have been
+// admitted by provided ingress controller.
+func (r *reconciler) clearAllRoutesStatusForIngressController(icName string) []error {
+	// List all routes.
+	errs := []error{}
+	start := time.Now()
+	routeList := &routev1.RouteList{}
+	routesCleared := 0
+	if err := r.client.List(context.TODO(), routeList); err != nil {
+		return append(errs, fmt.Errorf("failed to list all routes in order to clear route status for deployment %s: %w", icName, err))
+	}
+	// Clear status on the routes that belonged to icName.
+	for _, route := range routeList.Items {
+		if cleared, err := r.clearRouteStatus(&route, icName); err != nil {
```

Comment:
We did this micro-optimization in `clearRouteStatus`, might as well do it here:
```suggestion
	for i := range routeList.Items {
		if cleared, err := r.clearRouteStatus(&route[i], icName); err != nil {
```